Patch cherry picked from upstream:
https://github.com/libarchive/libarchive/commit/c34952a78284b505c88546a3d0a056182bf1c0da

Made small modifications to backport it to libarchive 3.4.1.

--- libarchive-3.4.1/libarchive/archive_read_support_format_lha.c	2020-01-21 12:48:41.570066685 +0000
+++ libarchive-3.4.1/libarchive/archive_read_support_format_lha.c	2020-01-21 12:48:25.016795623 +0000
@@ -1286,17 +1286,24 @@ lha_read_file_extended_header(struct arc
 				return (ARCHIVE_FATAL);
 			else {
 				/*
-				 * Convert directory delimiter from 0xFF
+				 * Convert directory delimiter from 0xFFFF
 				 * to '/' for local system.
 				 */
+				uint16_t dirSep;
+				uint16_t d = 1;
+				if (archive_be16dec(&d) == 1)
+				  dirSep = 0x2F00;
+				else
+				  dirSep = 0x002F;
+
 				/* UTF-16LE character */
 				uint16_t *utf16name = (uint16_t *)lha->dirname.s;
 				for (i = 0; i < lha->dirname.length / 2; i++) {
 					if (utf16name[i] == 0xFFFF)
-						utf16name[i] = L'/';
+						utf16name[i] = dirSep;
 				}
 				/* Is last character directory separator? */
-				if (utf16name[lha->dirname.length / 2 - 1] != L'/')
+				if (utf16name[lha->dirname.length / 2 - 1] != dirSep)
 					/* invalid directory data */
 					goto invalid;
 			}
